File: python/whilec.py
#!/usr/bin/env python3
"""WHILE 言語コンパイラのエントリポイント

    python whilec.py ../test/assign.while           # .wat を出力
    python whilec.py --stack ../test/assign.while   # 仮想スタック命令を表示
    python whilec.py --run ../test/assign.while     # 仮想機械で実行
    python whilec.py --ast ../test/assign.while     # AST を表示
"""

from __future__ import annotations

import argparse
import sys
from pathlib import Path

from emit_wasm import emit_wat
from interpret import interpret
from lexer import tokenize
from parser import parse
from syntax import string_of_stmt
from virtual_stack import compile_stack, format_code


def compile_source(src: str):
    tokens = tokenize(src)
    ast = parse(tokens)
    return ast, compile_stack(ast)


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description="WHILE to WebAssembly コンパイラ")
    ap.add_argument("source", help=".while ファイル")
    ap.add_argument("--ast", action="store_true", help="構文木を表示する")
    ap.add_argument("--stack", action="store_true", help="仮想スタック命令を表示する")
    ap.add_argument("--run", action="store_true", help="仮想スタック命令を実行する")
    ap.add_argument("-o", "--output", help="出力 .wat のパス")
    args = ap.parse_args(argv)

    path = Path(args.source)
    if not path.exists():
        print(f"ファイルがありません: {path}", file=sys.stderr)
        return 1

    src = path.read_text(encoding="utf-8")
    try:
        ast, code = compile_source(src)
    except Exception as e:
        print(f"error: {e}", file=sys.stderr)
        return 1

    if args.ast:
        print(string_of_stmt(ast))
        return 0

    if args.stack:
        print(format_code(code), end="")
        return 0

    if args.run:
        interpret(code)
        return 0

    wat = emit_wat(code)
    out = Path(args.output) if args.output else path.with_suffix(".wat")
    out.write_text(wat, encoding="utf-8")
    print(f"wrote {out}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: c/whilec.c
#include "emit_wasm.h"
#include "lexer.h"
#include "parser.h"
#include "syntax.h"
#include "util.h"
#include "virtual_stack.h"

#include <stdio.h>
#include <string.h>

static int usage(const char *argv0) {
    fprintf(stderr, "[usage] %s [--ast|--stack] [-o out.wat] file.while\n", argv0);
    return 1;
}

int main(int argc, char **argv) {
    init_console();
    int ast = 0, stack = 0;
    const char *outpath = NULL;
    const char *srcpath = NULL;
    for (int i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "--ast"))
            ast = 1;
        else if (!strcmp(argv[i], "--stack"))
            stack = 1;
        else if (!strcmp(argv[i], "-o") || !strcmp(argv[i], "--output")) {
            if (++i >= argc)
                return usage(argv[0]);
            outpath = argv[i];
        } else if (argv[i][0] == '-')
            return usage(argv[0]);
        else
            srcpath = argv[i];
    }
    if (!srcpath)
        return usage(argv[0]);

    char *src = xread_file(srcpath);
    Stmt *tree = parse(tokenize(src));
    clear_todo();
    Code code = compile_stack(tree);
    const char *todo = take_todo();
    if (todo) {
        fprintf(stderr, "error: %s\n", todo);
        return 1;
    }

    if (ast) {
        printf("%s\n", string_of_stmt(tree));
        return 0;
    }
    if (stack) {
        fputs(format_code(code), stdout);
        return 0;
    }

    char *wat = emit_wat(code);
    todo = take_todo();
    if (todo) {
        fprintf(stderr, "error: %s\n", todo);
        return 1;
    }

    char auto_out[512];
    if (!outpath) {
        snprintf(auto_out, sizeof auto_out, "%s", srcpath);
        char *dot = strrchr(auto_out, '.');
        if (dot)
            strcpy(dot, ".wat");
        else
            strcat(auto_out, ".wat");
        outpath = auto_out;
    }
    FILE *f = fopen(outpath, "w");
    if (!f)
        die("cannot write %s", outpath);
    fputs(wat, f);
    fclose(f);
    printf("wrote %s\n", outpath);
    return 0;
}
